### PR TITLE
🔬 Inspector: Add tests for PageSpeedPanel component edge cases

### PR DESCRIPTION
**Pull request by nilesh32236**
💡 Gap: `PageSpeedPanel.js` had missing coverage for the successful API scan paths (parsing returned metrics and suggestion lists), as well as complex edge cases like empty URL input blocks and network exceptions thrown during recursive polling.

🎯 Coverage:
- Validated that successful scans correctly parse and render metrics and call `onSuggestionsReady`.
- Validated early-return logic for empty URLs.
- Validated error boundary rendering when the `getPagespeedResults` recursive polling throws network exceptions.
- Enhanced robustness of test async execution by flushing microtasks and replacing `jest.runAllTimers` with interval-based timer advancement.

📊 Tools Used: Jest, React Testing Library, `npm run test -- --coverage`.

---
*PR created automatically by Jules for task [7670031541702799474](https://jules.google.com/task/7670031541702799474) started by @nilesh32236*

## Summary by CodeRabbit

* **Tests**
  * Improved coverage for PageSpeed polling behavior, including failures and exceptions.
  * Added validation for successful results and suggestion rendering.
  * Added coverage to prevent scans when no URL is provided.
  * Improved reliability of asynchronous timer and polling tests.

### CODE REVIEW

File: src/components/__tests__/PageSpeedPanel.test.js [hunk 1]
```diff
@@ -277,20 +277,148 @@ describe( 'PageSpeedPanel Component', () => {
...
+			expect(
+				screen.getByText( /PageSpeed scan is running/i )
+			).toBeInTheDocument();
+		} );
+
+		const consoleSpy = jest
+			.spyOn( console, 'error' )
+			.mockImplementation( () => {} );
+
+		getPagespeedResults.mockRejectedValueOnce( new Error( 'Network issue' ) );
```
Comment (nilesh32236) [inline src/components/__tests__/PageSpeedPanel.test.js]: 🔵 **MINOR**: ESLint/Prettier (via @wordpress/scripts) reports 3 errors on the added lines (328, 406, 408). `npx eslint` exits with code 1, so the required `npm run lint:js` CI gate in AGENTS.md will fail and block the merge. Verified locally: jest passes, eslint fails only on these three added lines.

> 💡 **How to fix:** Run `npm run lint:js:fix` (or `npx eslint --fix`) and commit the reformatted lines; re-run `npm run lint:js` before pushing.

```suggestion
getPagespeedResults.mockRejectedValueOnce(
			new Error( 'Network issue' )
		);
```

**Review by nilesh32236 — COMMENTED** [review]
## Executive Summary

**Purpose:** Adds 3 new Jest tests for PageSpeedPanel (polling exception catch-branch, successful scan result rendering + onSuggestionsReady propagation, and empty-URL early-return guard) and adds comments/extra timer advancement to an existing polling-failure test, plus a .jules/inspector.md note on the timer-flush technique.

**Risk:** 🟢 LOW — Test-only changes; no production code touched. All 16 tests pass. Only risk is CI lint:js failing on 3 prettier formatting errors in the added lines, plus misleading comments and a doc/code mismatch that could confuse future maintainers.

---

## MR Review Summary

The new tests genuinely improve coverage of PageSpeedPanel: they exercise the previously untested getPagespeedResults rejection path, verify the real score-gauge/vitals rendering and onSuggestionsReady callback on success, and lock in the empty-URL guard. All 16 tests pass locally. However the PR currently breaks `npm run lint:js` (3 prettier errors on added lines), and the edited polling-failure test picks up confusingly incorrect comments plus a redundant third timer-advancement phase that contradicts the documented approach.

**Ready to merge?** false

**Reasoning:** Tests are correct and pass, but CI verification will fail because eslint/prettier rejects 3 lines in the added code (the required gate in AGENTS.md runs npm run lint:js before merging). All blocking issues are trivial auto-fixable formatting problems.

### Strengths

- **src/components/__tests__/PageSpeedPanel.test.js:344** — The success-path test verifies real component behavior rather than mock wiring: it asserts the parsed score values (95/60/30), vitals display values, the category label mapping ('Performance', 'First Contentful Paint'), and the onSuggestionsReady propagation with the exact suggestions array. This is the strongest coverage added and closes the described gap.
- **src/components/__tests__/PageSpeedPanel.test.js:307** — The polling-exception test covers the previously untested catch branch in pollForResults (component line 208) and correctly mocks console.error since the component logs the caught error.
- **src/components/__tests__/PageSpeedPanel.test.js:412** — The empty-URL test validates the handleScan early-return guard (component line 227) — a real edge case that prevents a pointless API call.

File: src/components/__tests__/PageSpeedPanel.test.js [hunk 2]
```diff
@@ -277,20 +277,148 @@ describe( 'PageSpeedPanel Component', () => {
...
 			jest.runAllTimers();
 		} );
 		await act( async () => {} );
+
+		// Run a second time to resolve polling promise
 		await act( async () => {
 			jest.runAllTimers();
 		} );
 		await act( async () => {} );
+		// 1st timer tick calls poll() which gets status 'not_ready' (if mocked) or the result.
```
Comment (nilesh32236) [inline src/components/__tests__/PageSpeedPanel.test.js]: 🔵 **MINOR**: The added comments in the 'displays error when polling returns failure' test are factually wrong for this test: they claim 'we mocked the 1st getPagespeedResults to return the actual mockResult immediately', but the mocks here return `{status:'not_ready'}` and then a failure message — there is no mockResult in this test. These comments appear copy-pasted from the success test and will actively mislead anyone debugging why the polling logic behaves the way it does. Line 280 ('Let the effect execute') and line 286 are also vague/duplicated.

> 💡 **How to fix:** Delete the stale comment block (lines 291-296) and the vague 'Let the effect execute' line; keep only the original accurate comment describing the two poll phases (not_ready then failure).

```suggestion
Delete the stale comment block (lines 291-296) and the vague 'Let the effect execute' line; keep only the original accurate comment describing the two poll phases (not_ready then failure).
```

File: src/components/__tests__/PageSpeedPanel.test.js [hunk 3]
```diff
@@ -277,20 +277,148 @@ describe( 'PageSpeedPanel Component', () => {
...
+		// In this test, we mocked the 1st getPagespeedResults to return the actual mockResult immediately.
+		// However, handleScan polls recursively.
+		// Since we mocked getPagespeedResults to resolve immediately, the first timer
+		// tick will process the mockResult. However, getPagespeedResults is called inside the poll async function.
+		// We need to resolve all promises before running timers.
+		await act( async () => {
```
Comment (nilesh32236) [inline src/components/__tests__/PageSpeedPanel.test.js:283]: 🔵 **MINOR**: The added `jest.advanceTimersByTime(5000)` block is a redundant no-op: the two preceding runAllTimers phases already consumed both mocked getPagespeedResults responses, and the failure path calls stopPolling() which clears any pending timer, so nothing fires on the third advancement. Mixing three different timer-advancement strategies (runAllTimers x2, then advanceTimersByTime) in one test makes the timing logic far harder to reason about than the original two-phase version.

> 💡 **How to fix:** Remove the advanceTimersByTime(5000) block (lines 297-300) — the original two runAllTimers phases were sufficient and the test still passes without it (verified).

```suggestion
Remove the advanceTimersByTime(5000) block (lines 297-300) — the original two runAllTimers phases were sufficient and the test still passes without it (verified).
```

File: .jules/inspector.md [hunk 4]
```diff
@@ -31,3 +31,7 @@
 **Bug/Gap:** Tests triggering expected API failures were writing ugly `console.error` logs to the test runner output, confusing developers.
 **Root Cause:** The `DatabaseCleanup` component logs caught errors when the API fails using `console.error()`, which Jest mirrors directly to the CLI unless mocked.
 **Test Added:** Mocked `console.error` locally using `jest.spyOn(console, 'error').mockImplementation(() => {});` specifically for the exact test cases where errors are simulated, and correctly restored the mock via `consoleSpy.mockRestore();`.
+## 2026-08-01 - Test Timeout Issue with Async Polling Timers
```
Comment (nilesh32236) [inline .jules/inspector.md:35]: 🔵 **MINOR**: The new documentation entry claims the fix was 'swapping jest.runAllTimers() with controlled jest.advanceTimersByTime(interval)', but the committed code still uses jest.runAllTimers() in the modified polling-failure test (lines 282, 288) and the new exception test (line 330) — it merely added an extra advanceTimersByTime block alongside runAllTimers. The documented technique does not match the actual implementation, which misleads future contributors trying to reproduce or extend the fix.

> 💡 **How to fix:** Either update the entry to describe what was actually done (that advanceTimersByTime was used in the new success test while runAllTimers remains in the failure-path tests), or standardize the tests to consistently use the advanceTimersByTime approach described and re-verify they pass.

```suggestion
Either update the entry to describe what was actually done (that advanceTimersByTime was used in the new success test while runAllTimers remains in the failure-path tests), or standardize the tests to consistently use the advanceTimersByTime approach described and re-verify they pass.
```

File: src/components/__tests__/PageSpeedPanel.test.js [hunk 5]
```diff
@@ -277,20 +277,148 @@ describe( 'PageSpeedPanel Component', () => {
...
+			screen.getByRole( 'button', { name: /Run PageSpeed Scan/i } )
+		);
+
+		await waitFor( () => {
+			expect(
+				screen.getByText( /PageSpeed scan is running/i )
+			).toBeInTheDocument();
+		} );
+
+		const consoleSpy = jest
```
Comment (nilesh32236) [inline src/components/__tests__/PageSpeedPanel.test.js]: 🔵 **MINOR**: consoleSpy.mockRestore() at line 341 only executes if every assertion passes. If the waitFor times out or an earlier assertion throws, the console.error spy leaks into all subsequent tests: beforeEach only calls jest.clearAllMocks() (which clears call data but does NOT restore a spied-on implementation), so real errors would be silently swallowed and failures become very hard to diagnose. The same fragile pattern exists in the pre-existing 'displays error on scan exception' test, but the new test duplicates it. **[medium confidence]**

> 💡 **How to fix:** Restore the spy in afterEach (e.g. `afterEach( () => jest.restoreAllMocks() )`) or wrap the body in try/finally so the mock is always removed even on assertion failure.

```suggestion
afterEach( () => {
		jest.useRealTimers();
		jest.restoreAllMocks();
	} );
```

File: src/components/__tests__/PageSpeedPanel.test.js [hunk 6]
```diff
@@ -291,6 +292,126 @@ describe( 'PageSpeedPanel Component', () => {
...
```
Comment (nilesh32236) [inline src/components/__tests__/PageSpeedPanel.test.js]: 🔵 **MINOR**: ESLint no-unused-vars error: `const consoleSpy =` is assigned but never referenced in the new 'displays error when polling throws an exception' test (line 312). The same error exists at line 197 in the 'displays error on scan exception' test because this PR removed the only usage, `consoleSpy.mockRestore();`. `npx eslint` exits with code 1, so the required `npm run lint:js` CI gate in AGENTS.md and the webpack.yml workflow will fail and block the merge.

> 💡 **How to fix:** Drop the unused assignment and rely on the afterEach jest.restoreAllMocks() that this PR already adds to restore the spy; apply the same fix at line 197.

```suggestion
jest.spyOn( console, 'error' ).mockImplementation( () => {} );
```

File: src/components/__tests__/PageSpeedPanel.test.js [hunk 7]
```diff
@@ -291,6 +292,126 @@ describe( 'PageSpeedPanel Component', () => {
...
+			jest.runAllTimers();
+		} );
+		await act( async () => {} );
+
+		await waitFor( () => {
+			expect(
+				screen.getByText( 'PageSpeed scan failed.' )
+			).toBeInTheDocument();
+		} );
+
```
Comment (nilesh32236) [inline src/components/__tests__/PageSpeedPanel.test.js]: 🔵 **MINOR**: Prettier reports 'Delete ⏎' (prettier/prettier): there is a stray blank line between the closing `});` of the waitFor block and the test's closing `} );` in the 'displays error when polling throws an exception' test, which also contributes to the eslint failure.

> 💡 **How to fix:** Remove the empty line 330 so the waitFor closing and the test closing brace are adjacent.

File: .jules/inspector.md (hunk 4 above)
Comment (nilesh32236) [inline .jules/inspector.md:35]: 🔵 **MINOR**: Markdownlint MD022 (blanks-around-headings): the new '## 2026-08-01 - Test Timeout Issue with Async Polling Timers' heading has no blank line above (line 33 is the previous entry's 'Test Added:' line) or below (line 35 immediately follows). This is also the open CodeRabbit thread on this PR.

> 💡 **How to fix:** Add one blank line immediately before and after the heading on line 34.

File: update_inspector.sh [hunk 8]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
```
Comment (nilesh32236) [inline update_inspector.sh]: 🔵 **MINOR**: A one-off sed patch script is committed at the repo root purely to reword a single changelog line. It uses a brittle exact-match sed pattern (any edit to the doc text silently no-ops it) and becomes dead weight once this PR lands. **[medium confidence]**

> 💡 **How to fix:** Remove the script and apply the wording change directly to .jules/inspector.md, or if the Jules workflow needs it, move it under .jules/ and give it a clear purpose/comment.

**Review by nilesh32236 — COMMENTED** [review]
## Executive Summary

**Purpose:** Adds Jest coverage for PageSpeedPanel successful-scan metric/suggestion rendering, empty-URL guard, and network-exception polling paths, plus a test-timing reliability fix and an inspector.md changelog entry.

**Risk:** 🟢 LOW — Test-only changes to a UI component's test suite; no production code is modified and all 16 tests pass. The blocker is tooling: the PR introduces 3 ESLint errors that fail the required npm run lint:js CI gate, and a minor markdown lint violation. All issues are trivially auto-fixable.

---

## MR Review Summary

The new tests genuinely exercise the real rendering paths (metric parsing, vitals labels, onSuggestionsReady callback) rather than only asserting on mocks, and adding jest.restoreAllMocks() to afterEach is a solid reliability improvement. However, the PR fails the required eslint gate (3 errors on lines 197, 312, 330) and the committed docs/scripts contain minor inconsistencies, so it needs a small cleanup pass before merge.

**Ready to merge?** false

**Reasoning:** The lint:js CI gate fails with 3 eslint/prettier errors on the modified file, and the inspector.md changelog entry still contradicts the committed code. Fixes are small and mechanical.

### Strengths

- **src/components/__tests__/PageSpeedPanel.test.js:333** — The success-path test is genuinely valuable: it asserts that real parsed values (scores 95/60/30, vitals display values, 'Performance' label) are rendered and that onSuggestionsReady receives the exact suggestions array, verifying actual component behavior instead of just mock plumbing.
- **src/components/__tests__/PageSpeedPanel.test.js:56** — Adding jest.restoreAllMocks() to afterEach cleanly eliminates the console.error spy-leak risk for both the pre-existing scan-exception test and the new polling-exception test, which was previously only handled inline.
- **src/components/__tests__/PageSpeedPanel.test.js:280** — Replacing jest.runAllTimers() with paired jest.advanceTimersByTime(5000) + empty `await act(async () => {})` correctly models the recursive await-then-schedule polling loop and is the right pattern for this component.

### Issues

- 🔵 **MINOR:** `.jules/inspector.md:37` — The changelog entry claims the fix was 'consistently using controlled jest.advanceTimersByTime(interval) ... instead of jest.runAllTimers()', but the new test committed at PageSpeedPanel.test.js:321 still calls jest.runAllTimers(). The documentation contradicts the code it documents.
  > 💡 **How to fix:** Change the code at line 321 to `jest.advanceTimersByTime( 5000 );` (equivalent here since the rejected poll stops polling and schedules no further timers), or soften the doc wording to reflect that runAllTimers is used where only a single timer exists.
<details><summary>Show suggested fix</summary>

```suggestion
jest.advanceTimersByTime( 5000 );
```
</details>

File: src/components/__tests__/PageSpeedPanel.test.js [hunk 9]
```diff
@@ -291,6 +290,123 @@ describe( 'PageSpeedPanel Component', () => {
...
+		jest.spyOn( console, 'error' ).mockImplementation( () => {} );
+
+		getPagespeedResults.mockRejectedValueOnce(
+			new Error( 'Network issue' )
+		);
+
+		await act( async () => {
+			jest.runAllTimers();
```
Comment (nilesh32236) [inline src/components/__tests__/PageSpeedPanel.test.js]: 🔵 **MINOR**: The newly added 'displays error when polling throws an exception' test still uses `jest.runAllTimers()`, contradicting the changelog entry in .jules/inspector.md:39 which claims the fix was applied by 'consistently using controlled jest.advanceTimersByTime(interval) ... instead of jest.runAllTimers()'. The doc is now factually inaccurate about the committed code and the suite's convention is inconsistent (the two sibling polling tests use advanceTimersByTime). It happens to work here only because the mocked rejection terminates polling on the first tick, so no recursive scheduling is exercised.

> 💡 **How to fix:** Align the new test with the documented pattern by replacing `jest.runAllTimers()` with `await act(async () => { jest.advanceTimersByTime(5000); });` followed by `await act(async () => {});`, or adjust the inspector.md wording to state the pattern applies where recursive polling is exercised.

```suggestion
await act( async () => {
			jest.advanceTimersByTime( 5000 );
		} );
		await act( async () => {} );
```

File: src/components/__tests__/PageSpeedPanel.test.js [hunk 10]
```diff
@@ -291,6 +290,123 @@ describe( 'PageSpeedPanel Component', () => {
...
+		await act( async () => {} );
+
+		await waitFor( () => {
+			expect(
+				screen.getByText( 'PageSpeed scan failed.' )
+			).toBeInTheDocument();
+		} );
+	} );
+
+	it( 'displays scan results and calls onSuggestionsReady on success', async () => {
```
Comment (nilesh32236) [inline src/components/__tests__/PageSpeedPanel.test.js:328]: 🔵 **MINOR**: The success-path test mocks getPagespeedResults to return a ready result on the FIRST poll, so it never exercises the real-world recursive flow (not_ready on poll 1 → ready on poll 2) nor the negative branch where `response.data.suggestions` is falsy and onSuggestionsReady must NOT be called (PageSpeedPanel.js:201-206). This leaves the exact transition the PR title targets ('recursive polling succeeds') and its guard logic unverified end-to-end. **[medium confidence]**

> 💡 **How to fix:** Add a variant that returns `{ status: 'not_ready' }` on the first poll and the full result on the second, asserting results render only after the second advance; optionally add a success case without `suggestions` asserting onSuggestionsReady is not called.

```suggestion
Add a variant that returns `{ status: 'not_ready' }` on the first poll and the full result on the second, asserting results render only after the second advance; optionally add a success case without `suggestions` asserting onSuggestionsReady is not called.
```

**Review by nilesh32236 — COMMENTED** [review]
## Executive Summary

**Purpose:** Adds Jest test coverage for PageSpeedPanel edge cases (polling exceptions, successful result/suggestion parsing, empty-URL early return) and documents the async-polling timer fix in .jules/inspector.md.

**Risk:** 🟢 LOW — Test-only PR (one test file + one changelog doc). All 16 tests pass and `npx eslint` is clean; no production code or security surface is touched.

---

## MR Review Summary

Well-scoped, deterministic test additions that exercise real rendering logic (ScoreGauge, VitalRow, onSuggestionsReady wiring) rather than just mock plumbing. The PR also properly fixes the console.error spy leak via `afterEach(jest.restoreAllMocks())` and deletes the stray update_inspector.sh script. Remaining issues are minor documentation inconsistencies and a small coverage gap.

**Ready to merge?** true

**Reasoning:** Only minor, non-blocking issues remain: the inspector.md changelog overstates that runAllTimers was replaced consistently (one new test still uses it), and the success-path test short-circuits to a ready result without covering the recursive not_ready→success transition or the no-suggestions negative case. Neither affects mergeability.

### Strengths

- **src/components/__tests__/PageSpeedPanel.test.js:56** — Adding `jest.restoreAllMocks()` to afterEach replaces the brittle per-test `consoleSpy.mockRestore()` pattern, ensuring console.error spies can never leak into subsequent tests even when a waitFor times out or an assertion throws.
- **src/components/__tests__/PageSpeedPanel.test.js:328** — The success-path test asserts real DOM output (score values, vital display values, category/vital labels) and verifies `onSuggestionsReady` is invoked with the exact suggestions array — this validates actual parsing/rendering logic rather than only mock call plumbing.
- **src/components/__tests__/PageSpeedPanel.test.js:400** — The empty-URL test directly exercises the component's `if ( ! url ... ) return;` guard in handleScan, confirming the early-return blocks the API call as documented in the PR description.
- **src/components/__tests__/PageSpeedPanel.test.js:278** — Replacing `jest.runAllTimers()` with paired `jest.advanceTimersByTime(5000)` + `await act(async () => {})` correctly models the recursive poll loop (fire poll → flush microtask → schedule next poll), matching the root cause documented in inspector.md.

**Review by nilesh32236 — COMMENTED** [review]
## Executive Summary

**Purpose:** Adds Jest coverage for PageSpeedPanel successful-scan rendering (metrics, vitals, onSuggestionsReady), empty-URL blocking, and recursive polling failure/exception paths, while hardening async timer handling in the existing tests.

**Risk:** 🟢 LOW — Test-only changes: all 17 tests pass, eslint exits clean, and the previously reported console.error spy leak and runAllTimers/doc mismatches are fixed. No production code is touched.

---

## MR Review Summary

The PR delivers its stated goals: successful scans are now verified to render scores/vitals and call onSuggestionsReady, the empty-URL guard is covered, and polling exceptions are tested. The reviewer feedback from the prior round is substantively resolved — console.error spies are now restored via jest.restoreAllMocks(), the code consistently uses advanceTimersByTime wrapped in act() with microtask flushes, the stale docs are accurate, and eslint/prettier pass. Two minor items remain: the success-path test still returns a ready result on the first poll (not exercising the not_ready recursion), and the empty-URL test/guard misses whitespace-only URLs.

**Ready to merge?** true

**Reasoning:** All tests pass (17/17), the required npm run lint:js gate is green, and every previously reported issue except one low-severity coverage nit has been addressed. Remaining findings are minor and non-blocking for a test-only PR.

### Strengths

- **src/components/__tests__/PageSpeedPanel.test.js:56** — Adding jest.restoreAllMocks() to afterEach and dropping the assigned consoleSpy variable (jest.spyOn(console, 'error') without a binding) cleanly fixes the prior spy-leak and no-unused-vars problems: spies are now restored even when an assertion inside a test throws, and the mock is auto-restored between tests.
- **src/components/__tests__/PageSpeedPanel.test.js:400** — The 'displays scan results and does not call onSuggestionsReady if suggestions are falsy' test genuinely exercises the real-world recursive flow (first poll not_ready -> second poll ready) and asserts the intermediate non-rendered state with queryByText, which is the strongest coverage added in this PR.
- **src/components/__tests__/PageSpeedPanel.test.js:328** — The success-path test verifies actual parsed output (numeric scores, vitals display values, category label) rather than only asserting mock call counts, and confirms onSuggestionsReady is invoked with the suggestion payload — a meaningful end-to-end assertion of the component's render contract.
- **.jules/inspector.md:35** — The new changelog entry is now accurate: the committed code does use jest.advanceTimersByTime(5000) inside act() with empty await act(async () => {}) microtask flushes, matching the documented root cause about runAllTimers skipping intermediate promise resolutions. Markdownlint heading spacing is also correct.

### Issues

- 🔵 **MINOR:** `src/components/__tests__/PageSpeedPanel.test.js:353` — The success-path test still mocks getPagespeedResults to return a ready result on the FIRST poll, so it never exercises the not_ready -> ready recursive polling it is billed to validate; it only confirms a single immediate-success poll renders correctly. The recursion is covered only by the separate falsy-suggestions test. This is a coverage-quality nit, not a failing test. **[medium confidence]**
  > 💡 **How to fix:** Make the success test mirror real-world polling: mock the first getPagespeedResults call to return { success: true, data: { status: 'not_ready' } } and the second to return the ready mockResult, then advance timers twice (as the falsy test already does) before asserting rendered metrics.
- 🔵 **MINOR:** `src/components/__tests__/PageSpeedPanel.test.js:463` — The 'does not start scan if url is empty' test only covers the empty string. The component's guard is `if ( ! url || scanning || pending || submittingRef.current ) return;` (PageSpeedPanel.js:227) — a whitespace-only URL like '   ' is truthy, so it passes the guard and would call queuePagespeedScan against an invalid URL. The new edge-case test silently codifies this gap. **[medium confidence]**
  > 💡 **How to fix:** Extend the test to also assert a whitespace-only URL does not start a scan (it currently would fail), and consider trimming the URL in handleScan: `if ( ! url?.trim() || ... ) return;` in PageSpeedPanel.js.
